Sandbox/src/Game/NpcController.cpp

```diff
@@ -0,0 +1,234 @@
+#include "NpcController.h"
+
+NpcController::NpcController(int npcIndex, Point spawnPoint, std::shared_ptr<Map> map, std::shared_ptr<Player> player)
+	: mNpcIndex(npcIndex), mSpawnPoint(spawnPoint), mMap(map), mPlayer(player)
+{
+	mSpawnTimer = 0.0f;
+	mNewPosition = false;
+
+	SpawnNpc();
+	mNpcState = NpcState::Wandering;
+}
+
+void NpcController::Invoke(ICommand* command)
+{
+	if (command->CanExecute())
+		command->Execute();
+}
+
+void NpcController::SpawnNpc()
+{
+	mNpc = NpcFactory::Instance().GetNpc(mNpcIndex);
+	mNpc->MapPosition(mSpawnPoint);
+	mSpawnTimer = 0.0f;
+
+	mDeathTimer = 0.6f;
+
+	mWanderTimer = 0.0f;
+
+	if (std::shared_ptr<NpcFighter> npcFighter = std::dynamic_pointer_cast<NpcFighter>(mNpc))
+	{
+		mAttackDelay = 0.0f;
+		mSpawnTimer = npcFighter->RespawnTime();
+	}
+}
+
+void NpcController::Wander()
+{
+	if (!mNpc->InCombat() && mNpc->CurrentPath().empty())
+	{
+		if (mWanderTimer <= 0.0f)
+		{
+			if (100.0f * Random::Float() <= 33.0f)
+			{
+				static Point direction[4] = { {1, 0}, {-1, 0}, {0, 1}, {0, -1} };
+				Point pos = mNpc->MapPosition() + direction[rand() % 4];
+
+				int dx = mSpawnPoint.x - pos.x;
+				int dy = mSpawnPoint.y - pos.y;
+
+				if (abs(dx) < 11 && abs(dy) < 11)
+				{
+					if (mMap->TileWalkable(pos))
+					{
+						mNpc->MoveTo(pos);
+					}
+				}
+			}
+
+			if (rand() % 2)
+				mWanderTimer = 1.0f;
+		}
+		else
+			mWanderTimer -= Square::Timer::Instance().DeltaTime();
+	}
+}
+
+bool NpcController::MoveTo(Point p)
+{
+	if (mMap->TileWalkable(p))
+	{
+		mNpc->MoveTo(p);
+		return true;
+	}
+
+	return false;
+}
+
+bool NpcController::MoveInRange()
```

Comment:
this is a pretty gnarly looking function
if we want to reduce the indentations we can combine the first three conditionals into one (and tag variable names for better readability if we want) or move things into helper functions

Reply:
I agree this function could definitely use some cleaning up lmao